cfd_protocol/src/lib.rs

```diff
@@ -558,31 +621,161 @@ mod tests {
             };
 
             let mut tx_refund = self.inner;
-            self.lock_output_descriptor
+            self.commit_descriptor
                 .satisfy(&mut tx_refund.input[0], satisfier)?;
 
             Ok(tx_refund)
         }
     }
 
+    #[derive(Debug, Clone)]
+    struct PunishTransaction {
+        inner: Transaction,
+    }
+
+    impl PunishTransaction {
+        fn new(
+            commit_tx: &CommitTransaction,
+            address: &Address,
+            amount: Amount,
+            encisg: EcdsaAdaptorSignature,
+            sk: SecretKey,
+            (revocation_them_sk, revocation_them_pk): (SecretKey, PublicKey),
+            publishing_them_pk: PublicKey,
+            revoked_commit_tx: Transaction,
+        ) -> Result<Self> {
+            // CommitTransaction has only one input
+            let input = revoked_commit_tx.input[0].clone();
```

Comment:
With `Itertools`, you could encode this invariant using `into_iter` and `exactly_one`: https://docs.rs/itertools/0.10.1/itertools/trait.Itertools.html#method.exactly_one